### ISSUE TITLE
MissingOverride: ignore generated method

### ISSUE THREAD

**Issue by pacbeckh** [google/error-prone#4065]
MissingOverride generates warnings for methods that have `@Generated` annotation.

In https://github.com/google/error-prone/pull/1968 there was a fix done by skipping classes that have the annotation.
However, if just parts of the class are generated, as with Lombok, the class level `@Generated` annotation is missing.

I would expect methods with `@Generated` to be ignored/skipped.

### Minimal Reproducible Example

```java
package org.example;

import lombok.Generated;

public class AnyClassDeLombok implements AnyInterface {
  private String anything;

  @Generated
  public String getAnything() {
    return this.anything;
  }

}
```

A full reproducable example is available at https://github.com/pacbeckh/error-prone-missing-override-issue,  run `mvn clean install` 

<details>
<summary>Logs</summary>

```sh
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project errorproneissue: Compilation failure: Compilation failure: 
[ERROR] /Users/pacovanbeckhoven/hexagon/error-prone-missing-override-issue/src/main/java/org/example/AnyClassDeLombok.java:[12,17] [MissingOverride] getAnything implements method in AnyInterface; expected @Override
[ERROR]     (see https://errorprone.info/bugpattern/MissingOverride)
[ERROR]   Did you mean '@Override @Generated'?
[ERROR] /Users/pacovanbeckhoven/hexagon/error-prone-missing-override-issue/src/main/java/org/example/AnyClass.java:[8,18] [MissingOverride] getAnything implements method in AnyInterface; expected @Override
[ERROR]     (see https://errorprone.info/bugpattern/MissingOverride)
[ERROR]   Did you mean 'private String @Override anything;'?
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```
</details>

### Setup

- Operating system  `MacOS Ventura`.
- Java version `17.0.4`).
- Error Prone version `2.21.1`.

**Comment by Stephan202** [comment]
While looking into (the different but related) PicnicSupermarket/error-prone-support#716, I noticed that Lombok [by default](https://projectlombok.org/features/configuration) adds `@SuppressWarnings("all")` to generated code. The Immutables library does the same.

So an alternative works-out-of-the-box-for-all-checks solution could be the following:
```diff
diff --git a/check_api/src/main/java/com/google/errorprone/SuppressionInfo.java b/check_api/src/main/java/com/google/errorprone/SuppressionInfo.java
index cb23c08dea..3b891b9cc5 100644
--- a/check_api/src/main/java/com/google/errorprone/SuppressionInfo.java
+++ b/check_api/src/main/java/com/google/errorprone/SuppressionInfo.java
@@ -85,7 +85,8 @@ public class SuppressionInfo {
       return SuppressedState.SUPPRESSED;
     }
     if (suppressible.supportsSuppressWarnings()
-        && !Collections.disjoint(suppressible.allNames(), suppressWarningsStrings)) {
+        && (suppressWarningsStrings.contains("all")
+            || !Collections.disjoint(suppressible.allNames(), suppressWarningsStrings))) {
       return SuppressedState.SUPPRESSED;
     }
     if (suppressible.suppressedByAnyOf(customSuppressions, state)) {
diff --git a/check_api/src/main/java/com/google/errorprone/bugpatterns/BugChecker.java b/check_api/src/main/java/com/google/errorprone/bugpatterns/BugChecker.java
index 4693b70ef8..f6485134a0 100644
--- a/check_api/src/main/java/com/google/errorprone/bugpatterns/BugChecker.java
+++ b/check_api/src/main/java/com/google/errorprone/bugpatterns/BugChecker.java
@@ -99,6 +99,7 @@ import java.io.Serializable;
 import java.lang.annotation.Annotation;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 import java.util.Set;
 import java.util.function.BiPredicate;
@@ -269,8 +270,12 @@ public abstract class BugChecker implements Suppressible, Serializable {
   }
 
   private boolean isSuppressed(SuppressWarnings suppression) {
-    return suppression != null
-        && !Collections.disjoint(Arrays.asList(suppression.value()), allNames());
+    if (suppression == null) {
+      return false;
+    }
+
+    List<String> suppressions = Arrays.asList(suppression.value());
+    return suppressions.contains("all") || !Collections.disjoint(suppressions, allNames());
   }
 
   /**
```

If the Error Prone maintainers think that this is an acceptable solution then I wouldn't mind writing some unit tests for this and opening a PR.

**Comment by cushon** [comment]
> If the Error Prone maintainers think that this is an acceptable solution then I wouldn't mind writing some unit tests for this and opening a PR.

That SGTM overall. I think your proposal handles that, but it would be good to have test coverage to double-check it doesn't apply to 'undisableable' checks.

I have some mixed feelings about `@SuppressWarnings("all")`.

* Our internal style is to prefer specific suppressions with a comment explaining them (e.g. `@SuppressWarnings("unchecked") // safe covariant cast`).

* But just ignoring `@SuppressWarnings("all")` doesn't seem very friendly, and recognizing it might address some of the issues with the interaction between Error Prone and the Lombok language.

* One possibility might be to have yet another flag to configure whether `all` is supported, or yet another check to discourage its use.

**Comment by Stephan202** [comment]
> I have some mixed feelings about `@SuppressWarnings("all")`.

Agree. Both suggestions SGTM, though a separate check is perhaps a bit more versatile, as it would also cover suppressions targeted at other static analysis tools.

W.r.t. opening a PR: I've added it to my TODO list for the coming days; need to first focus on a few other topics. (The modified methods and surrounding `SuppressionInfo` and `BugChecker` classes don't yet have (directly) associated tests. I suppose creating such classes is a bit nicer than modifying some "random" other class.)

**Comment by pacbeckh** [comment]
I think this would solve a lot of issues, as this is not the only check that is conflicting with Lombok.
Especially since the `@Generated` annotations are not always added on e.g. class level.

Thanks @Stephan202 and @cushon for the quick response 👌

**Comment by Stephan202** [comment]
Practical question for @cushon :)

Context: if made `@VisibleForTesting`, writing tests for the (effectively deprecated) method in `BugChecker` isn't so hard, as can be seen on [this branch](https://github.com/google/error-prone/compare/master...PicnicSupermarket:error-prone:sschroevers/suppress-warnings-all). _Part_ of the `SuppressionInfo` changes could also be tested this way if its constructor is made `@VisibleForTesting`.

But: _complete_ testing of the modified methods, without changing method visibility, requires constructing suitable `Symbol` instances. For this I see two options:
1. Using some boilerplate code to compile dummy code like in [`SignaturesTest`](https://github.com/google/error-prone/blob/a992ae7c821020282d30c5169add0c74fc52e6ea/check_api/src/test/java/com/google/errorprone/util/SignaturesTest.java#L48-L90).
2. Using a `CompilationTestHelper`, which would _require defining the test class in `core` rather than `check_api`_.

I think option (2) will yield the most robust and maintainable code, and this is the approach also taken for `SuggestedFixes` and `SuggestedFixesTest`, but splitting code across modules like this is not optimal. So before I present a PR that requires rework: any opinions on the preferred approach? :)

**Comment by cushon** [comment]
Thanks for working on this @Stephan202!

I'm happy to be pragmatic about either of the options you described. If writing the test with `CompilationTestHelper` is cleaner, I think it's OK to define the test in `core`. I would also be OK with yet another copy of the boilerplate in `SignaturesTest` to set up a javac invocation.

**Comment by Stephan202** [comment]
Thanks for the pointers @cushon! Eventually I settled on option (2). I filed #4076 with a proposal.
